src/main/java/org/opensearch/neuralsearch/processor/util/ProcessorUtils.java

```diff
@@ -130,37 +133,50 @@ public static void removeTargetFieldFromSource(final Map<String, Object> sourceA
             parentMap = currentParentMapWithChild.v1();
             key = currentParentMapWithChild.v2();
 
-            @SuppressWarnings("unchecked")
-            Map<String, Object> innerMap = (Map<String, Object>) parentMap.get(key);
+            Map<String, Object> innerMap = castToMap(parentMap.get(key));
 
             if (innerMap != null && innerMap.isEmpty()) {
                 parentMap.remove(key);
             }
         }
     }
 
+    public static Optional<Object> getValueFromSource(final Map<String, Object> sourceAsMap, final String targetField) {
+        return getValueFromSource(sourceAsMap, targetField, -1);
+    }
+
     /**
      * Returns the mapping associated with a path to a value, otherwise
      * returns an empty optional when it encounters a dead end.
-     * <hr>
      * When the targetField has the form (key[.key]) it will iterate through
      * the map to see if a mapping exists.
+     * When a List is encountered during traversal, the index parameter is used to select
+     * the appropriate element from the list. If the selected element is a map, traversal continues with
+     * that map using the next key in the path.
      *
      * @param sourceAsMap The Source map (a map of maps) to iterate through
      * @param targetField The path to take to get the desired mapping
+     * @param index       the index to use when a list is encountered during traversal; if list processing is not needed,
+     *                    -1 is passed in
      * @return A possible result within an optional
      */
-    public static Optional<Object> getValueFromSource(final Map<String, Object> sourceAsMap, final String targetField) {
+    public static Optional<Object> getValueFromSource(final Map<String, Object> sourceAsMap, final String targetField, int index) {
         String[] keys = targetField.split("\\.");
-        Optional<Object> currentValue = Optional.of(sourceAsMap);
+        Optional<Object> currentValue = Optional.ofNullable(sourceAsMap);
 
         for (String key : keys) {
             currentValue = currentValue.flatMap(value -> {
-                if (!(value instanceof Map<?, ?>)) {
-                    return Optional.empty();
+                if (value instanceof ArrayList<?> && index != -1) {
+                    Object listValue = (castToObjectList(value)).get(index);
+                    if (listValue instanceof Map) {
+                        Map<String, Object> currentMap = castToMap(listValue);
+                        return Optional.ofNullable(currentMap.get(key));
+                    }
+                } else if (value instanceof Map<?, ?>) {
+                    Map<String, Object> currentMap = castToMap(value);
+                    return Optional.ofNullable(currentMap.get(key));
                 }
-                Map<String, Object> currentMap = (Map<String, Object>) value;
-                return Optional.ofNullable(currentMap.get(key));
+                return Optional.empty();
```

Comment:
Nit. For me better to put this under else block to make sure every block should return itself without missing any condition.
```
else {
  Optional.empty();
}
```